llvm/lib/Analysis/ValueTracking.cpp

```diff
@@ -3142,6 +3159,9 @@ static bool isKnownNonEqual(const Value *V1, const Value *V2, unsigned Depth,
     };
   }
 
+  if (isNonEqualSELECT(V1, V2, Depth, Q) || isNonEqualSELECT(V2, V1, Depth, Q))
```

Comment:
This is going to do duplicate work for the two selects + same condition case.

Reply:
For this case, `isNonEqualSelect` will return true and second call won't happen at all. I checked this with print statements. I am compiling `opt` in`Debug` mode and this is still true.

Or did I misunderstand your statement?

Reply:
What I meant here is that if isNonEqualSelect returns `false`, then the second isNonEqualSelect call with swapped operands is also going to return false. Not super important though.

Reply:
Oh, ok. I can do a separate PR later and refactor this and other similar calls like `if(isNonEqualMul(V1, V2, Depth, Q) || isNonEqualMul(V2, V1, Depth, Q))`.